pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java

```
@@ -290,63 +318,63 @@ public void run() {
-                    sendOutputMessage(srcRecord, result1.getResult());
-                } else {
-                    if (instanceConfig.getFunctionDetails().getAutoAck()) {
-                        // the function doesn't produce any result or the user doesn't want the result.
-                        srcRecord.ack();
-                    }
-                }
-                // increment total successfully processed
-                stats.incrTotalProcessedSuccessfully();
+    @SuppressWarnings("rawtypes")
+	private void processResult(Record srcRecord, JavaExecutionResult result) throws Exception {
+    	
+    	if (result.getUserException() != null) {
+    		log.warn("Encountered exception when processing message {}",
+                    srcRecord, result.getUserException());
+            stats.incrUserExceptions(result.getUserException());
+            throw result.getUserException();
+    	} 
+    	
+    	if (result.getSystemException() != null) {
+    		log.warn("Encountered exception when processing message {}",
+                    srcRecord, result.getSystemException());
+            stats.incrSysExceptions(result.getSystemException());
+            throw result.getSystemException();
+    	}
+    	
+    	if (result.getResult() == null) {
+    		if (instanceConfig.getFunctionDetails().getAutoAck()) {
+                // the function doesn't produce any result or the user doesn't want the result.
+                srcRecord.ack();
             }
-        });
+    	} else { 
+    		
+    		try {
+				Object output = (result.getResult() instanceof CompletableFuture) ?
+					((CompletableFuture)result.getResult()).get() : result.getResult();
+				
+				sendOutputMessage(srcRecord, output);
+				
+				if (instanceConfig.getFunctionDetails().getAutoAck()) {
+					srcRecord.ack();
+				}
+				stats.incrTotalProcessedSuccessfully();
+				
+			} catch (InterruptedException | ExecutionException e) {
+				log.warn("Encountered exception when processing message {}",
```

please handle InterruptedException

---

@eolivelli Just to clarify, you want to call `Thread.currentThread().interrupt()` when we encounter an `InterruptedException`?  Why? and why only for that exception an not the `ExecutionException` ?

---

because it is the standard Java practice.
when you catch a InterruptedException the "interrupted" flag is reset on the Thread and you have to set it again in order to let the information bubble up to the caller.

> why only for that exception an not the ExecutionException
because ExecutionException is not the special (and annoying) InterruptedException